File: forensics/oracle/probes/probe_name.c
/*
 * probe_name.c — oracle probe for dns_name_* semantics (bind9-rs courts).
 *
 * Reads one test case per line from stdin.  Each line is a raw name in
 * masterfile text form.  Writes one result line per input to stdout:
 *
 *   OK <formatted> <wire-hex> <countlabels> <length>
 *   ERR <result-code>
 *
 * formatted = dns_name_format() output; wire-hex = uncompressed wire bytes
 * in lowercase hex; countlabels/length = dns_name_countlabels()/
 * dns_name_length().  The origin is the root, so relative names become
 * absolute (the same behavior bind9-rs must reproduce).
 *
 * This probe links against the pinned BIND tree's libdns/libisc.  It exists
 * ONLY on the oracle side (spec §2).
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include <isc/buffer.h>
#include <isc/mem.h>
#include <isc/result.h>
#include <isc/util.h>

#include <dns/compress.h>
#include <dns/fixedname.h>
#include <dns/name.h>
#include <dns/result.h>

static isc_mem_t *mctx = NULL;
static dns_compress_t cctx;

static void
emit_name(const dns_name_t *name) {
	char text[DNS_NAME_MAXTEXT + 1];
	unsigned char wire[DNS_NAME_MAXWIRE];
	isc_buffer_t tb;
	isc_buffer_t wb;
	isc_region_t r;
	char hex[DNS_NAME_MAXWIRE * 2 + 1];
	unsigned int i;
	isc_result_t result;

	isc_buffer_init(&tb, text, sizeof(text));
	result = dns_name_totext(name, false, &tb);
	if (result != ISC_R_SUCCESS) {
		printf("TOTEXT-ERR %s\n", isc_result_totext(result));
		return;
	}
	text[isc_buffer_usedlength(&tb)] = '\0';

	isc_buffer_init(&wb, wire, sizeof(wire));
	dns_compress_rollback(&cctx, 0);
	result = dns_name_towire(name, &cctx, &wb, NULL);
	if (result != ISC_R_SUCCESS) {
		printf("WIRE-ERR %s\n", isc_result_totext(result));
		return;
	}
	isc_buffer_usedregion(&wb, &r);
	for (i = 0; i < r.length; i++) {
		sprintf(hex + i * 2, "%02x", r.base[i]);
	}
	hex[r.length * 2] = '\0';

	printf("OK %s %s %u %u\n", text, hex, dns_name_countlabels(name),
	       name->length);
}

int
main(int argc, char **argv) {
	char line[4096];
	isc_result_t result;
	dns_fixedname_t fname;
	dns_name_t *name;
	dns_fixedname_t forigin;
	dns_name_t *origin;
	isc_buffer_t source;

	UNUSED(argc);
	UNUSED(argv);

	isc_mem_create(&mctx);
	dns_compress_init(&cctx, mctx, 0);
	dns_fixedname_init(&fname);
	name = dns_fixedname_name(&fname);

	/* Origin = root. */
	dns_fixedname_init(&forigin);
	origin = dns_fixedname_name(&forigin);
	isc_buffer_init(&source, ".", 1);
	isc_buffer_add(&source, 1);
	result = dns_name_fromtext(origin, &source, NULL, 0, NULL);
	if (result != ISC_R_SUCCESS) {
		fprintf(stderr, "cannot build root origin: %s\n",
			isc_result_totext(result));
		return 1;
	}

	while (fgets(line, sizeof(line), stdin) != NULL) {
		size_t len = strlen(line);
		while (len > 0 && (line[len - 1] == '\n' || line[len - 1] == '\r')) {
			line[--len] = '\0';
		}
		if (len == 0) {
			continue;
		}
		isc_buffer_init(&source, line, (unsigned int)len);
		isc_buffer_add(&source, (unsigned int)len);

		dns_name_reset(name);
		result = dns_name_fromtext(name, &source, origin, 0, NULL);
		if (result != ISC_R_SUCCESS) {
			printf("ERR %s\n", isc_result_totext(result));
			continue;
		}
		emit_name(name);
	}

	isc_mem_destroy(&mctx);
	return 0;
}

File: bind9-rs-forensics/src/bin/probe-name.rs
//! `probe-name` — the bind9-rs side of the `CORE-NAME-TEXT-*` courts.
//!
//! Mirrors `forensics/oracle/probes/probe_name.c` byte-for-byte: reads one
//! raw name per stdin line, emits
//!
//! ```text
//! OK <formatted> <wire-hex> <countlabels> <length>
//! ERR <result-text>
//! ```
//!
//! where the result-text is BIND's `isc_result_totext` string (see
//! `bind9-core::error`).  The oracle probe links real libdns; this probe
//! uses bind9-core.  Byte-identical output is the parity claim; anything
//! else becomes a residual.

use bind9_rs_core::error::Error;
use bind9_rs_core::name::Name;
use std::io::{BufRead, Write};

fn main() {
    let root = Name::root();
    let stdin = std::io::stdin();
    let mut out = std::io::stdout().lock();
    for line in stdin.lock().lines() {
        let line = line.unwrap_or_default();
        if line.is_empty() {
            continue;
        }
        match Name::from_text(&line, Some(&root)) {
            Ok(name) => {
                let text = name.to_text();
                let mut wire = Vec::new();
                if bind9_rs_core::name::wire::to_wire_uncompressed(&name, &mut wire).is_err() {
                    let _ = writeln!(out, "WIRE-ERR internal");
                    continue;
                }
                let hex: String = wire.iter().map(|b| format!("{b:02x}")).collect();
                let _ = writeln!(
                    out,
                    "OK {} {} {} {}",
                    text,
                    hex,
                    name.label_count(),
                    name.wire_len_full()
                );
            }
            Err(e) => {
                let _ = writeln!(out, "ERR {}", err_text(&e));
            }
        }
    }
}

/// The BIND `isc_result_totext` string for an error.
fn err_text(e: &Error) -> String {
    e.to_string()
}
